pkgs/development/python-modules/jaxlib/default.nix

```diff
@@ -108,7 +108,8 @@ let
       libjpeg_turbo
       numpy
       pkgs.flatbuffers
-      pkgs.protobuf
+      # See https://github.com/google/jax/issues/9534.
+      pkgs.protobuf3_12
```

Comment:
Ahh that's a great catch... I wonder if this would be problematic at all. It seems that protobuf isn't included in `propagatedBuildInputs` anywhere though, so I don't think they'd conflict in PYTHONPATH... 

I does give me some https://github.com/NixOS/nixpkgs/issues/156767 flashbacks though. 

I didn't try patching since that seemed like a dangerous rabbit hole, but maybe I'll give it a shot.

Reply:
Ok I just tested this and everything seems to work fine.
```
❯ NIX_PATH=.. nix-shell -p python3Packages.jaxlib python3Packages.tensorflow                                  

[nix-shell:~/dev/nixpkgs]$ echo $PYTHONPATH | tr ':' '\n' | grep protobuf
/nix/store/j7jbqfagpar0h2xvklzxrjp74443fqaf-python3.9-protobuf-3.19.3/lib/python3.9/site-packages

[nix-shell:~/dev/nixpkgs]$ echo $LD_LIBRARY_PATH | tr ':' '\n' | grep protobuf

[nix-shell:~/dev/nixpkgs]$ echo $PATH | tr ':' '\n' | grep protobuf
/nix/store/kyajg85hhf8anb8rk0j5dpg9ma9yyyxk-protobuf-3.19.3/bin

[nix-shell:~/dev/nixpkgs]$ python
Python 3.9.10 (main, Jan 13 2022, 23:32:03) 
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import jaxlib
>>> import tensorflow
>>> 
```
I think because protobuf isn't in `propagatedBuildInputs` in jaxlib.

Does this look ok to you @ndl? Is there anything I might be missing?

Reply:
Just to clarify - I'm not concerned about Python `protobuf` versions conflict but about the underlying `.so` libraries conflict. 

The resulting Python `.so` extensions for `jaxlib` contain the reference to `libprotobuf.so`:

```
ldd /nix/store/lqv8zg74ha8zw33kkr4ybh05zd27qiin-python3.9-jaxlib-0.1.75/lib/python3.9/site-packages/jaxlib/xla_extension.so | grep protobuf
```
should show smth like

```
libprotobuf.so.30 => /nix/store/kyajg85hhf8anb8rk0j5dpg9ma9yyyxk-protobuf-3.19.3/lib/libprotobuf.so.30
```

In addition, if you run the following command:
```
LD_DEBUG=files nix-shell -p "with (import ./.{}); python3.withPackages(ps: [ ps.jaxlib ps.jax ])" --run "python3 -c 'import jax.numpy as jnp; jnp.random()'" 2>&1 | grep protobuf.so
```
you'll get the output that looks like this:
```
calling init: /nix/store/kyajg85hhf8anb8rk0j5dpg9ma9yyyxk-protobuf-3.19.3/lib/libprotobuf.so.30
```

This demonstrates that importing and using `jax` (which, in turn, will import and use `jaxlib`) will cause `libprotobuf.so` to be loaded into the process.

Now, if there's another package that also uses `libprotobuf.so` (either Python `protobuf` implementation or any other package that links to `libprotobuf.so`), which `libprotobuf.so` gets imported depends on the import order of the packages. If they all import the same `libprotobuf.so` - that doesn't matter, but if the versions are different - you might get some "interesting" (and somewhat hard to debug) issues.

Now, I don't want to block your PR based on this - fully hypothetical ATM - scenario, just explaining what I meant in my comments above and what types of problematic scenarios _might_ surface :-)

Reply:
Ah ok, thanks for explaining! My mental model was that things would be ok as long as each .so doesn't link against two versions of protobuf, causing a conflict in the symbol table (ie everything's ok as long as ldd only has one protobuf). But is this still problematic if A.so links against protobuf1 and B.so links against protobuf2?

Reply:
I guess these are the sorts of cases where dynamic linking is so annoying...

Reply:
> is this still problematic if A.so links against protobuf1 and B.so links against protobuf2?

In general - "it depends" :-) I'm by no means an expert in this topic, but IIUC depending on whether you use dynamic loading via `dlopen` [with some fancy flags](https://stackoverflow.com/questions/36545396/dlopen-doesnt-respect-rtld-local), play with [symbols visibility](https://stackoverflow.com/questions/435352/limiting-visibility-of-symbols-when-linking-shared-libraries) and / or use [symbols versioning](https://man7.org/conf/lca2006/shared_libraries/slide19a.html), use different [SONAME](https://en.wikipedia.org/wiki/Soname), etc - you can get all sorts of interesting combinations and results.

Speaking of our particular case: `protobuf` [does seem to use SONAME for versioning](https://github.com/protocolbuffers/protobuf/blob/master/src/Makefile.am#L21), so different dynamic libraries versions will get different library names and will be loaded separately. It also uses [visibility control](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/port_def.inc#L437) so, in principle, it should export only its APIs (= reducing the surface for symbols conflicts). OTOH it doesn't look like there's any symbols versioning policy, so I'd still expect the conflicts of symbols are possible when two different versions of the library are loaded. I'm not 100% sure if the errors [like this](https://stackoverflow.com/questions/56801994/how-to-use-shared-library-with-two-different-versions-indirectly-in-my-project) are related to symbols conflicts or smth else, though - needs deeper investigation.

... I guess that's a long and convoluted way of saying "I still don't know whether it will work" :-) Simple test case of importing two different Python extensions linked with different protobuf shared libraries versions "works" (= doesn't crash), but I also haven't tried doing anything meaningful there.

I guess that if patching the code or switching to static linking is not an option, you can try going ahead and using your current PR "as is" and see if it causes any issues in practice?

Reply:
Well you sure sound like an expert to me 👍  At least I'm learning a lot here, which is great! Thank you so much for your detailed response!

Let me take a stab at patching. If that proves to be sufficiently complicated I vote that we merge this as is and try to fix in a future PR. Hopefully the SONAME versioning is sufficient to avoid conflicts. At minimum I think we should be compatible with `python3Packages.tensorflow`. From what I understand there's no way for us to eliminate the possibility of symbol collisions entirely since users could always import some arbitrary other python package that happens to depend load an so with a conflicting protobuf version anyway. So the best we can do is to try to be good nixpkgs citizens and follow the latest `protobuf`.

Reply:
Implemented the patch solution in https://github.com/NixOS/nixpkgs/pull/159099/commits/f3753388ad7291ee20f9bd07c25b9a66ed62269a. Not as messy as I expected!